db/db_iter.cc

```diff
@@ -1590,6 +1594,7 @@ void DBIter::SeekToFirst() {
     Seek(*iterate_lower_bound_);
     return;
   }
+  PERF_COUNTER_ADD(iter_seek_to_first_count, 1);
```

Comment:
I'm unsure if we should only count `iter_seek_to_first_count` when there is no `iterator_lower_bound`, similar for `iter_seek_to_last_count`. From user's perspective, SeekToFirst() was called but not counted.

Reply:
Yeah - this is one concern. The other one I tried to avoid as you see is the double counting. The underlying design issue here is whether we want to have a separate counter for SeekToFirst/Last/ VS count them under Seek. If NO, then we solve both the double counting and user semantics (probably that's also why we have `PERF_CPU_TIMER_GUARD(iter_seek_cpu_nanos, clock_);` but not `PERF_CPU_TIMER_GUARD(iter_seek__to_first_cpu_nanos, clock_);`. Let me double check on the user need as I don't have a preference here.

Reply:
I did not realize we use `iter_seek_cpu_nanos` for all Seek* APIs. Given that, using a single Seek counter seems fine.

 I think we can also count Seek* APIs separately and avoid double counting by doing `PERF_COUNTER_ADD(*, -1);`.

Reply:
> Let me double check on the user need as

Confirmed - one counter is fine for that particular user need  (also fine for general case I believe as mentioned above) so will change to 1 seek counter for all

> and avoid double counting by doing PERF_COUNTER_ADD(*, -1);

Do you mean, for example,  something like below:

```
void DBIter::SeekToFirst() {
  PERF_COUNTER_ADD(seek_to_first_count, 1);
  if (iterate_lower_bound_ != nullptr) {
    Seek(*iterate_lower_bound_);
      PERF_COUNTER_ADD(seek_to_first_count, -1);
    return;
  }
```

Reply:
If yes, I think it's indeed a simpler hack than passing parameter around to indicate whether some count in Seek() will be double-counted (i.e, will be offset later)!

Though, I still can't help wonder why we have to think about hacks like that, as the downside of that does exist e.g, PERF_COUNTER_ADD is supposed to be "increase";  PERF_COUNTER_ADD(seek, -1); after calling seek isn't that obvious to understand. 

I believe that's because "SeekToFirst()" and "Seek()" aren't implemented as "independently" as we want in order to count their API calls correctly (compared with Prev()/Next()/Seek()). To be more specific, if there was an internal function like InternalSeek() that encapsulates the shared code path between Seek() and SeekToFirst(), we can avoid awkness by below

```
void DBIter::SeekToFirst() {
  PERF_COUNTER_ADD(seek_to_first_count, 1);
  if (iterate_lower_bound_ != nullptr) {
    InternalSeek(*iterate_lower_bound_);
    return;
  }
  
  
void DBIter::Seek(const Slice& target) {
  PERF_COUNTER_ADD(seek_count, 1);
  InternalSeek(target);
}
```

What's your thoughts?

Reply:
> Do you mean, for example, something like below: ..

Yeah.

> if there was an internal function like InternalSeek() ..

Good point, I think we use SeekImpl() under Seek() in several places in the codebase. I'm fine with either approach (aggregate every Seek* in a single seek counter or count separately).